Sources/_CShims/process_shims.c

```diff
@@ -0,0 +1,298 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+#include "include/_CShimsTargetConditionals.h"
+#include "include/process_shims.h"
+#include <errno.h>
+#include <signal.h>
+#include <sys/wait.h>
+
+#if __has_include(<grp.h>)
+#include <grp.h>
+#endif
+
+int _was_process_exited(int status) {
+    return WIFEXITED(status);
+}
+
+int _get_exit_code(int status) {
+    return WEXITSTATUS(status);
+}
+
+int _was_process_signaled(int status) {
+    return WIFSIGNALED(status);
+}
+
+int _get_signal_code(int status) {
+    return WTERMSIG(status);
+}
+
+#if TARGET_OS_LINUX
+#include <stdio.h>
+
+int _shims_snprintf(
+    char * _Nonnull str,
+    int len,
+    const char * _Nonnull format,
+    char * _Nonnull str1,
+    char * _Nonnull str2
+) {
+    return snprintf(str, len, format, str1, str2);
+}
+#endif
+
+// MARK: - Darwin (posix_spawn)
+#if TARGET_OS_MAC
+
+int _subprocess_spawn(
+    pid_t  * _Nonnull  pid,
+    const char  * _Nonnull  exec_path,
+    const posix_spawn_file_actions_t _Nullable * _Nonnull file_actions,
+    const posix_spawnattr_t _Nullable * _Nonnull spawn_attrs,
+    char * _Nullable const args[_Nonnull],
+    char * _Nullable const env[_Nullable],
+    uid_t * _Nullable uid,
+    gid_t * _Nullable gid,
+    int number_of_sgroups, const gid_t * _Nullable sgroups,
+    int create_session
+) {
+    int require_pre_fork = uid != NULL ||
+        gid != NULL ||
+        number_of_sgroups > 0 ||
+    create_session > 0;
+
+    if (require_pre_fork != 0) {
+        pid_t childPid = fork();
+        if (childPid != 0) {
+            *pid = childPid;
+            return childPid < 0 ? errno : 0;
+        }
+
+        if (uid != NULL) {
+            if (setuid(*uid) != 0) {
+                return errno;
+            }
+        }
+
+        if (gid != NULL) {
+            if (setgid(*gid) != 0) {
+                return errno;
+            }
+        }
+
+        if (number_of_sgroups > 0 && sgroups != NULL) {
+            if (setgroups(number_of_sgroups, sgroups) != 0) {
+                return errno;
+            }
+        }
+
+        if (create_session != 0) {
+            (void)setsid();
+        }
+    }
+
+    // Set POSIX_SPAWN_SETEXEC if we already forked
+    if (require_pre_fork) {
+        short flags = 0;
+        int rc = posix_spawnattr_getflags(spawn_attrs, &flags);
+        if (rc != 0) {
+            return rc;
+        }
+
+        rc = posix_spawnattr_setflags(
+            (posix_spawnattr_t *)spawn_attrs, flags | POSIX_SPAWN_SETEXEC);
+        if (rc != 0) {
+            return rc;
+        }
+    }
+
+    // Spawn
+    return posix_spawn(pid, exec_path, file_actions, spawn_attrs, args, env);
```

Comment:
While `posix_spawn` is fine in some circumstances, it is not always ideal. In some cases one needs to replace the current process with `execv` instead of launching a new one, for example the `swift run` command in SwiftPM works exactly this way. Is there anything in the current `Subprocess` API that would allow that? If not, is there anything in the API that would prevent us from adding it in the future?

Reply:
@MaxDesiatov  I think `POSIX_SPAWN_SETEXEC` makes `posix_spawn` essentially `execve` on steroids.